lib/runtime/storage/storagediff.go

```diff
@@ -0,0 +1,292 @@
+	if childTrieChanges != nil {
+		value, deleted := childTrieChanges.get(key)
+		return value, deleted
+	}
+
+	return nil, false
+}
+
+// upsertChild inserts or updates a value associated with a key within a
+// specific child trie. If the child trie or the key was previously marked for
+// deletion, this marking is reversed, and the value is updated.
+func (cs *storageDiff) upsertChild(keyToChild, key string, value []byte) {
+	if cs == nil {
+		return
+	}
+
+	// If we previously deleted this child trie we have to undo that deletion
+	if cs.deletes[keyToChild] {
+		delete(cs.deletes, keyToChild)
+	}
+
+	childChanges := cs.childChangeSet[keyToChild]
+	if childChanges == nil {
+		childChanges = newStorageDiff()
+	}
+
+	childChanges.upserts[key] = value
+	cs.childChangeSet[keyToChild] = childChanges
+}
+
+// deleteFromChild marks a key for deletion within a specific child trie.
+func (cs *storageDiff) deleteFromChild(keyToChild, key string) {
+	if cs == nil {
+		return
+	}
+
+	childChanges := cs.childChangeSet[keyToChild]
+	if childChanges == nil {
+		childChanges = newStorageDiff()
+	}
+
+	childChanges.delete(key)
+	cs.childChangeSet[keyToChild] = childChanges
+}
+
+// snapshot creates a deep copy of the current change set, including all upserts,
+// deletions, and child trie change sets.
+func (cs *storageDiff) snapshot() *storageDiff {
+	if cs == nil {
+		panic("Trying to create snapshot from nil change set")
+	}
+
+	childChangeSetCopy := make(map[string]*storageDiff)
+	for k, v := range cs.childChangeSet {
+		childChangeSetCopy[k] = v.snapshot()
+	}
+
+	return &storageDiff{
+		upserts:        maps.Clone(cs.upserts),
+		deletes:        maps.Clone(cs.deletes),
+		childChangeSet: childChangeSetCopy,
+	}
+}
+
+// applyToTrie applies all accumulated changes in the change set to the
+// provided trie. This includes insertions, deletions, and modifications in both
+// the main trie and child tries.
+// In case of errors during the application of changes, the method will panic
+func (cs *storageDiff) applyToTrie(t *trie.Trie) {
+	if cs == nil {
+		panic("trying to apply nil change set")
+	}
+
+	// Apply trie upserts
+	for k, v := range cs.upserts {
+		err := t.Put([]byte(k), v)
+		if err != nil {
+			panic("Error applying upserts changes to trie")
+		}
+	}
+
+	// Apply child trie upserts
+	for childKeyString, childChangeSet := range cs.childChangeSet {
+		childKey := []byte(childKeyString)
+
+		for k, v := range childChangeSet.upserts {
+			err := t.PutIntoChild(childKey, []byte(k), v)
+			if err != nil {
+				panic("Error applying child trie changes to trie")
+			}
+		}
+
+		for k := range childChangeSet.deletes {
+			err := t.ClearFromChild(childKey, []byte(k))
+			if err != nil {
+				if !strings.Contains(err.Error(), trie.ErrChildTrieDoesNotExist.Error()) {
+					panic("Error applying child trie keys deletion to trie")
+				}
+			}
+		}
+	}
+
+	// Apply trie deletions
+	for k := range cs.deletes {
+		key := []byte(k)
+		child, _ := t.GetChild(key)
```

Comment:
why ignore the error here?

Reply:
Usually the error we could have when we are getting a Child is that the child doesn't exists. I'm checking if `child != nil` for that so I think it's covered.
Another way to solve it is checking the error and ignoring it if it is `ErrChildTrieDoesNotExist` 

What do you think is the best way?

Reply:
Hey, as we talked [I added a TODO](https://github.com/ChainSafe/gossamer/pull/3777/commits/fdca5bf66379d720b6e324626ce2010ee1e95e6f) to discuss it later